services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AutoProvisioner.java

```diff
@@ -0,0 +1,386 @@
+        // to start(). This results in an exception in the factory's connect() method.
+        synchronized (this) {
+            if (connectionAttempt == null) {
+                connectionAttempt = downstreamSenderFactory.connect().map(connection -> {
+                    LOG.info("connected to AMQP network");
+                    if (vertx == null && connection != null) {
+                        vertx = connection.getVertx();
+                    }
+                    return connection;
+                }).recover(t -> {
+                    LOG.warn("failed to connect to AMQP network", t);
+                    return Future.failedFuture(t);
+                });
+            }
+        }
+
+        return connectionAttempt.mapEmpty();
+    }
+
+    @Override
+    public final Future<Void> stop() {
+        final Promise<Void> result = Promise.promise();
+        downstreamSenderFactory.disconnect(result);
+        return result.future();
+    }
+
+    /**
+     * Sets the vert.x instance.
+     *
+     * @param vertx The vert.x instance.
+     * @throws NullPointerException if vert.x is {@code null}.
+     */
+    @Autowired
+    public final void setVertx(final Vertx vertx) {
+        this.vertx = Objects.requireNonNull(vertx);
+    }
+
+    /**
+     * Sets the factory to use for creating a client for the AMQP Messaging Network.
+     *
+     * @param factory The factory.
+     * @throws NullPointerException if the factory is {@code null}.
+     */
+    @Autowired
+    public final void setDownstreamSenderFactory(final DownstreamSenderFactory factory) {
+        this.downstreamSenderFactory = Objects.requireNonNull(factory);
+    }
+
+    /**
+     * Sets the service to use for checking existence of tenants.
+     * <p>
+     * If not set, tenant existence will not be verified.
+     *
+     * @param tenantInformationService The tenant information service.
+     * @throws NullPointerException if service is {@code null};
+     */
+    @Autowired(required = false)
+    public final void setTenantInformationService(final TenantInformationService tenantInformationService) {
+        this.tenantInformationService = Objects.requireNonNull(tenantInformationService);
+        LOG.info("using {}", tenantInformationService);
+    }
+
+    /**
+     * Sets the OpenTracing {@code Tracer} to use for tracking the processing
+     * of messages published by devices across Hono's components.
+     * <p>
+     * If not set explicitly, the {@code NoopTracer} from OpenTracing will
+     * be used.
+     *
+     * @param opentracingTracer The tracer.
+     * @throws NullPointerException if the opentracingTracer is {@code null}.
+     */
+    @Autowired(required = false)
+    public final void setTracer(final Tracer opentracingTracer) {
+        LOG.info("using OpenTracing Tracer implementation [{}]", opentracingTracer.getClass().getName());
```

Comment:
What will happen to `opentracingTracer.getClass().getName()` when `opentracingTracer` is _null_? The _null_ check in the next line should be done first.